loadtest/control/simulcontroller/utils.go

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package simulcontroller
+
+import (
+	"errors"
+	"math/rand"
+)
+
+// pickAction randomly selects an action from a slice of userAction with
+// probability proportional to the action's frequency.
+func pickAction(actions []userAction) (*userAction, error) {
```

Comment:
Just to confirm: this frequency is the opposite of the runFrequency we have in simplecontroller right ? runFrequency denotes run this once in x times, and this `frequency` is just a weight of how frequently to run the action.

In that case, I would suggest to change the individual frequencies to be within [0,1]. Where 0 means it will never be selected, and 1 means maximum chance of selection. And make the total sum to be 1. The disadvantage is that adding new actions will need to shift the weights of other actions. But picking an action becomes super-simple because it just becomes similar to picking a node in consistent hashing where you just see where in the pie of [0,1] your number has fallen.

Right now, users can do 100, 1000, 5000. Basically there's no upper limit. Making the range from [0,1] should make things simple.

Thoughts ?

Reply:
> Just to confirm: this frequency is the opposite of the runFrequency we have in simplecontroller right ? runFrequency denotes run this once in x times, and this `frequency` is just a weight of how frequently to run the action.
> 

Yes this is more like an actual frequency. The higher the value, the higher the chance of picking it. The `runFrequency` of the `SimpleController` is really more of a period than a frequency (if we take 1 cycle/loop iteration to be our unit of time).

> In that case, I would suggest to change the individual frequencies to be within [0,1]. Where 0 means it will never be selected, and 1 means maximum chance of selection. And make the total sum to be 1. The disadvantage is that adding new actions will need to shift the weights of other actions. 
> 
> Right now, users can do 100, 1000, 5000. Basically there's no upper limit. Making the range from [0,1] should make things simple.
> 
> Thoughts ?

That's a fair point. Percentages are probably even easier to look at (total sum to 100).
But how we present data to the user is not necessarily how we deal with it internally.

I've looked at numerous implementations for the weighted random selection and most of them (at least the simple linear versions) use integers, I am guessing mainly to avoid possible rounding errors of floating point operations. 
Also, being able to be independent from the value of the total sum I think makes things simpler. As you pointed out, changing or adding an action in a slice of many actions means recalculating everything if we want to make the sum converge.

My idea is that while these frequencies/probabilities can be changed that's not something the user of the `load-test` program should do. I think there are better tools that we can provide that could affect the probability of an action without the need to change raw numbers.

> But picking an action becomes super-simple because it just becomes similar to picking a node in consistent hashing where you just see where in the pie of [0,1] your number has fallen.

Not very familiar with hashing techniques. Also not sure how using floating point would simplify the internal algorithm. Could you give an example?

Reply:
Yeah .. it's not a user knob though, which is good.

> Not very familiar with hashing techniques. Also not sure how using floating point would simplify the internal algorithm.

It's not really about the floating point, rather having a fixed bound. Take a look here: https://tom-e-white.com/2007/11/consistent-hashing.html. The idea is you put nodes in a sorted map, and then while getting the node, return the first value that you get from that point in the map.

Actually, scratch that. There was a subtle point that I missed. In consistent hashing, it wraps around to the earlier nodes because it's a circle. We don't want that here. Never mind then. :man_facepalming: